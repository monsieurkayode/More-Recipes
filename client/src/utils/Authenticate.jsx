import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'proptypes';

/**
 * Authentication HOC
 *
 * @param {JSX} ComposedComponent
 *
 * @returns {JSX} JSX
 */
export default function (ComposedComponent) {
  /**
 * @summary - Authenticate class declaration
 * @class Authenticate
 * @extends {Component}
 */
  class Authenticate extends Component {
    componentWillMount() {
      if (!this.props.isAuthenticated) {
        this.props.history.push('/signin');
      }
    }

    componentWillUpdate(nextProps) {
      if (!nextProps.isAuthenticated) {
        this.props.history.push('/');
      }
    }

    render() {
      return (
        <ComposedComponent {...this.props} />
      );
    }
  }

  const mapStateToProps = ({ signinState }) => ({
    user: signinState.user,
    isAuthenticated: signinState.isAuthenticated
  });

  Authenticate.propTypes = {
    history: PropTypes.shape({
      push: PropTypes.func.isRequired
    }).isRequired,
    isAuthenticated: PropTypes.bool.isRequired,
  };
  return connect(mapStateToProps)(Authenticate);
}
