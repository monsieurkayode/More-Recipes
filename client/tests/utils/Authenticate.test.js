/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import Authenticate from '../../src/utils/Authenticate';
import Footer from '../../src/components/footer';

const setup = (isAuthenticated) => {
  const props = {
    isAuthenticated,
    history: {
      push: jest.fn()
    }
  };

  const store = mockStore({
    signinState: {
      user: {
        username: 'user'
      }
    },
    isAuthenticated
  });


  const Component = () => <Footer />;

  Component.displayName = 'FooterComponent';

  const RequireAuthComponent = Authenticate(Component);

  const wrapper = mount(
    <Provider store={store}>
      <RequireAuthComponent {...props} />
    </Provider>
  );

  return { wrapper };
};

const { wrapper } = setup(true);

describe('Authenticate', () => {
  it('should return a component', () => {
    expect(wrapper.find('Connect(Authenticate)')).toHaveLength(1);
  });

  it('should render passed component if user is authenticated',
    (done) => {
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find('FooterComponent')).toHaveLength(1);
      done();
    });
});
