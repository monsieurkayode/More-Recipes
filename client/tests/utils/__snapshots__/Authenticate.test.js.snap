// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticate should render passed component if user is authenticated 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Authenticate)
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/signin",
            ],
          ],
        },
      }
    }
    isAuthenticated={true}
  >
    <Authenticate
      dispatch={[Function]}
      history={
        Object {
          "push": [MockFunction] {
            "calls": Array [
              Array [
                "/signin",
              ],
            ],
          },
        }
      }
      user={
        Object {
          "username": "user",
        }
      }
    >
      <FooterComponent
        dispatch={[Function]}
        history={
          Object {
            "push": [MockFunction] {
              "calls": Array [
                Array [
                  "/signin",
                ],
              ],
            },
          }
        }
        user={
          Object {
            "username": "user",
          }
        }
      >
        <Footer>
          <footer
            className="page-footer blue-grey darken-2"
          >
            <div
              className="row"
            >
              <div
                className="col l3 social-icon-container"
              >
                <ul>
                  <li>
                    <div>
                      <span>
                        <i
                          className="fa fa-2x fa-github teal-text"
                        />
                      </span>
                    </div>
                  </li>
                  <li>
                    <div>
                      <span>
                        <i
                          className="fa fa-2x fa-facebook teal-text"
                        />
                      </span>
                    </div>
                  </li>
                  <li>
                    <div>
                      <span>
                        <i
                          className="fa fa-2x fa-twitter teal-text"
                        />
                      </span>
                    </div>
                  </li>
                  <li>
                    <div>
                      <span>
                        <i
                          className="fa fa-2x fa-google teal-text"
                        />
                      </span>
                    </div>
                  </li>
                </ul>
              </div>
            </div>
            <div
              className="footer-copyright"
            >
              <div
                className="container"
              >
                Â© Andela 2018
                <a
                  className="grey-text text-lighten-4 right"
                  href="/"
                >
                  Created by Kayode Adeola
                </a>
              </div>
            </div>
          </footer>
        </Footer>
      </FooterComponent>
    </Authenticate>
  </Connect(Authenticate)>
</Provider>
`;
